tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/UnitTestRunnerRunSingleTestIntegration3_8.cs

```diff
@@ -0,0 +1,114 @@
+    ParameterTypeNames = ["Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.TestMethod", "System.Collections.Generic.IDictionary`2[System.String,System.Object]", "Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging.IMessageLogger"],
+    MinimumVersion = "14.0.0",
+    MaximumVersion = "14.*.*",
+    IntegrationName = MsTestIntegration.IntegrationName)]
+[Browsable(false)]
+[EditorBrowsable(EditorBrowsableState.Never)]
+// ReSharper disable once InconsistentNaming
+public static class UnitTestRunnerRunSingleTestIntegration3_8
+{
+    internal static CallTargetReturn<TReturn?> OnMethodEnd<TTarget, TReturn>(TTarget instance, TReturn? returnValue, Exception? exception, in CallTargetState state)
+    {
+        if (instance is null || !MsTestIntegration.IsEnabled)
+        {
+            return new CallTargetReturn<TReturn?>(returnValue);
+        }
+
+        var methodInfoCacheItem = MsTestIntegration.IsTestMethodRunnableThreadLocal.Value;
+        MsTestIntegration.IsTestMethodRunnableThreadLocal.Value = null;
+        if (methodInfoCacheItem is not null && returnValue is IList { Count: > 0 } lstResults)
+        {
+            foreach (var unitTestResultObject in lstResults)
+            {
+                if (unitTestResultObject != null &&
+                    unitTestResultObject.TryDuckCast<TestResultStruct3_8>(out var unitTestResult) &&
+                    methodInfoCacheItem.TestMethodInfo.TryDuckCast<ITestMethod>(out var testMethod))
+                {
+                    Common.Log.Debug("UnitTestRunner.RunSingleTest() call target interception: {Class}.{Name}", testMethod.TestClassName, testMethod.TestMethodName);
+
+                    if (unitTestResult.Outcome is UnitTestOutcome.Inconclusive or UnitTestOutcome.NotRunnable or UnitTestOutcome.Unknown)
+                    {
+                        if (!MsTestIntegration.ShouldSkip(testMethod, out _, out _))
+                        {
+                            // This instrumentation catches all tests being ignored
+                            MsTestIntegration.OnMethodBegin(testMethod, instance.GetType(), isRetry: false)?.Close(TestStatus.Skip, TimeSpan.Zero, unitTestResult.IgnoreReason);
+                        }
+                    }
+                    else if (unitTestResult.Outcome is UnitTestOutcome.Error or UnitTestOutcome.Failed)
+                    {
+                        if (methodInfoCacheItem.TestMethodInfo.TryDuckCast<ITestMethodInfo>(out var testMethodInfo))
+                        {
+                            // We need to check if the test is failing because a Class initialization error
+                            if (testMethodInfo.Parent?.Instance.TryDuckCast<ClassInfoInitializationExceptionStruct>(out var classInfoInitializationExceptionStruct) == true)
+                            {
+                                if (classInfoInitializationExceptionStruct.ClassInitializationException is { } classInitializationException &&
```

Comment:
Is it expected that duck typing to `ClassInfoInitializationExceptionStruct` could succeed, but `Exception` is null? 🤔

Reply:
yes because here `ClassInfoInitializationExceptionStruct` is extracted from a `TestClassInfo` so it can have an exception or not. `TestClassInfo` has a lot of properties regarding the class of the test method.